### PR TITLE
add status field to response in RPC methods `tx`, `EXPERIMENTAL_tx_status`, `broadcast_tx_commit`, `broadcast_tx_async`

### PR DESCRIPTION
**Pull request by telezhnaya**
This PR is a part of RPC tx methods redesign https://docs.google.com/document/d/1jGuXzBlMAGQENsUpy5x5Xd7huCLOd8k6tyMOEE41Rro/edit?usp=sharing
Addressing point 3 in https://github.com/near/nearcore/issues/9542 and start working on https://github.com/near/nearcore/issues/6837

Changes:
- New field `status` appeared in the response of RPC methods `tx`, `EXPERIMENTAL_tx_status`, `broadcast_tx_commit`, `broadcast_tx_async`
- Added some comments with the explanations

In https://github.com/near/nearcore/issues/6837, Illia suggested to have the structure
```rust
enum BroadcastWait {
 /// Returns right away.
 None,
 /// Waits only for inclusion into the block.
 Inclusion,
 /// Waits until block with inclusion is finalized.
 InclusionFinal,
 /// Waits until all non-refund receipts are executed.
 Executed,
 /// Waits until all non-refund receipts are executed and the last of the blocks is final.
 ExecutedFinal,
 /// Waits until everything has executed and is final.
 Final,
 }
```

I've renamed it to `TxExecutionStatus` and simplified it a little by dropping all refund options:
1. We may add them later without breaking backwards compatibility
2. To support them, we need to have the access to `receipt` (it's the method `get_tx_execution_status`). We have there only `execution_outcome` by default. We created special logic to be able not to retrieve the receipts (it's the only difference between `tx` and `EXPERIMENTAL_tx_status`). I have a feeling (please correct me here if I'm wrong) that retrieving corresponding receipt may slow down the execution that we tried to optimise.
BTW, maybe the data from `execution_outcome` is enough (we have there `gas_burnt` and `tokens_burnt`). @jakmeier suggested the condition `outcome.tokens_burnt == 0 && outcome.status == Success`. Unfortunately, we need to work with any status. But maybe the first part (`outcome.tokens_burnt == 0`) is enough to say that it could be only refund receipt. I need more input here.

### CODE REVIEW

**Comment by telezhnaya** [comment]
@wacban @jakmeier adding you back :)
I can't mark this PR as "ready for the review", it's still too raw.
Though I need your attention here. Could you please have a look at the general approach? Any of your input will be appreciated.

In the meantime, I'll think about tests. For now, I just roughly fixed the existing ones

File: chain/client/src/view_client.rs
Comment (jakmeier) [inline chain/client/src/view_client.rs:1182]: I think this PR introduced it: https://github.com/near/nearcore/pull/7895

File: core/primitives/src/views.rs
Comment (jakmeier) [inline core/primitives/src/views.rs:1658]: Oh, this is returned in the response? Interesting, I thought it will be part of the request, where the user specifies which finality guarantees they need.

Both approaches seem possible. From a user perspective, I imagine it is easier to ask for a specific finality and wait for the answer, as otherwise they have to keep polling until the desired finality is reached.
On the other hand, if we include the enum in the request, we have to do the polling on the RPC node which means more work on the server.

File: core/primitives/src/views.rs
Comment (jakmeier) [inline core/primitives/src/views.rs:1670]: For a field in a response, this enum variant is a bit akward IMO. It says all receipts were executed on some chain but it doesn't even guarantee that the transaction was included in a final block.

Idea: What if the response contains two fields?

```rust
inclusion: No/NotFinal/Final,
execution: No/NotFinal/Final,
```

This approach is not as nice from a strong typing point of view, since it allows encoding invalid states such as `inclusion: No, execution: Final`. But I think it's easier to use than having a single enum that has all valid combinations like this:

```rust
enum TxExecutionStatus {
    None,
    InclusionNonFinal_NotExecuted, // Inclusion
    InclusionNonFinal_ExecutedNonFinal,  // Executed (or Inclusion)
    InclusionFinal_NotExcecuted,  // Inclusion
    InclusionFinal_ExcecutedNonFinal, // Executed (or Inclusion)
    InclusionFinal_ExcecutedFinal, // Final
```

In comments how they map to your current variants.
You can probably find better names than mine to make the meaning clearer. :) But still, I find the two fields approach simpler from an API user perspective.

But yeah, I count 6 valid combinations, where the two-fields approach allow for 9 combinations. Not sure what's the best approach, just writing down all my thoughts :shrug:

File: core/primitives/src/views.rs
Comment (telezhnaya) [reply core/primitives/src/views.rs:1658]: I was also thinking about which approach is better.
I decided to move forward with returning status as fast as possible, with the finality info in the response.
And, when I'll work on `broadcast_tx`, I have a plan to add the same struct as an optional argument to the request.

Comment (jakmeier) [reply core/primitives/src/views.rs:1658]: I think "optional for the request, always included in the response" makes sense. As long as a user asking for `Inclusion` only gets a response where `Inclusion` or strictly stronger finality is guaranteed, and it is well documented, I think it should be okay.

Comment (jakmeier) [reply core/primitives/src/views.rs:1658]: Oh and yes, starting with just the field in the response first makes sense to me.

File: core/primitives/src/views.rs
Comment (telezhnaya) [reply core/primitives/src/views.rs:1670]: >  It says all receipts were executed on some chain but it doesn't even guarantee that the transaction was included in a final block.

I am sorry, I wasn't clear enough in these comments, I'll improve it. Current options go from the softest to the strictest. It means, if you received `Executed` response, the requirements for `InclusionFinal` were also satisfied.

File: core/primitives/src/views.rs
Comment (nikurt) [inline core/primitives/src/views.rs:1678]: Thinking about the corner cases from a point of view of a user:
* Submit a transaction
* Transaction gets included in a block in a fork
* See that the transaction is included in a block
* Wait for 10s
* See that the transaction is included in a block
* Wait for 10s
* .. repeat forever

Should we help the user in this case?
IIUC, this infinite loop is not possible with the execution of receipts, because the outgoing receipts of the parent block **must be** included in the child block.
Inclusion of transactions is different, because recording a transaction on-chain is not a mandatory action.

A counter-argument would be that a transaction may never be included in any block by chunk producers, and we can't possibly verify this condition.

Comment (telezhnaya) [reply core/primitives/src/views.rs:1678]: After we merge this, I'll start working on adding the same struct to the request to all the methods touched in this PR. It should solve this issue at least partially.

Thank you for the review!
I've addressed all the other comments and added the notes to CHANGELOG.md, please have a look again

**Review by nikurt — APPROVED** [review]
:+1: :+1: :+1:

**Pull request merged**
